src/datahike/db/utils.cljc

```diff
@@ -204,10 +204,15 @@
 
 (defn distinct-datoms [db current-datoms history-datoms]
   (if  (dbi/-keep-history? db)
-    (concat (filter #(or (no-history? db (:a %))
-                         (multival? db (:a %)))
-                    current-datoms)
-            history-datoms)
+    (let [current-datoms (filterv (fn [datom]
+                                    (let [a (:a datom)]
+                                      (or (no-history? db a)
+                                          (multival? db a))))
+                                  current-datoms)
+          current-datom-set (set current-datoms)]
+      (into current-datoms
```

Comment:
If I understand this correctly this transducer acts on `current-datoms` which is a persistent vector. This would mean you have to eagerly realize the full datom result set, while before this was lazy. Is this correct? It would be good to keep operations that act on large chunks of the database lazy, I think.

Reply:
I pushed some code but I am working on a better solution.

Reply:
OK, I tried to implement an approach where we assume that the two datoms sequences to merge are sorted and then try to merge the sorted sequences. If that would have worked, we would have a lazy sequence with distinct datoms without having to store all the current datoms in memory at a given time. An attempt is found here:

https://github.com/replikativ/datahike/compare/main...jonasseglare:datahike:distinct-history-datoms-dev?expand=1

But I was not able to get it to work. So I think that the code that as pushed is the laziest I can come up with for now.

Reply:
I like the merge function, I have thought about this approach in the past as well. Why does it not work?